safe_api.js

```diff
@@ -4,233 +4,295 @@ window.safeAPI = {};
+			.then(_ => safeApp.refreshContainersPermissions(this.handle))
+			.then(_ => console.log('Authorised, connected, refreshed permissions.'));
 		}
 
 		getContainersPermissions() {
-			return this.handlePromise.then((appHandle) => safeApp.getContainersPermissions(appHandle)); // Promise<Array<ContainerPerms>>
+			return this.then(_ => safeApp.getContainersPermissions(this.handle)); // Promise<Array<ContainerPerms>>
 		}
 
 		getOwnContainer() {
-			var homeContainerMD = new MutableData(this);
-			homeContainerMD.handlePromise = this.handlePromise.then((appHandle) => safeApp.getOwnContainer(appHandle));
-			return homeContainerMD;
+			return new MutableData(_ => safeApp.getOwnContainer(this.handle), this.root);
 		}
 
 		newPermissionsSet() {
-			var newSet = new PermissionsSet();
-			newSet.handlePromise = this.handlePromise.then((appHandle) => safeMutableData.newPermissionSet(appHandle));
-			return newSet;
+			return new PermissionsSet(_ => safeMutableData.newPermissionSet(this.handle), this.root);
 		}
 
 		newMutation() {
-			var newMut = new Mutation();
-			newMut.handlePromise = this.handlePromise.then((appHandle) => safeMutableData.newMutation(appHandle));
-			return newMut;
+			return new Mutation(_ => safeMutableData.newMutation(this.handle), this.root);
 		}
 
-		quickRandomPublic(data) {
+		quickRandomPublic(data, name, desc) {
 			//  looks like this does not get changed?
 			const typeTag = 15000;
 
-			var publicMD = new MutableData(this);
-			console.log('new publicMD: ', publicMD);
-			this.then(_ => safeMutableData.newRandomPublic(this.handleLocal, typeTag))
-			.then((mdata) => {
-				console.log('Random MD handle: ', mdata);
-				publicMD.mdHandle = mdata;
-				return safeMutableData.quickSetup(mdata, data);
-			})
-			.then(_ => {
-				console.log('MD handle was setup: ', publicMD.mdHandle);
-				return publicMD.mdHandle;
-			});
+			let publicMD = new MutableData(_ => safeMutableData.newRandomPublic(this.handle, typeTag), this.root);
+			publicMD.quickSetup(data, name, desc);
+			publicMD.then(_ => console.log('Successfully created and setup random public.'));
 			return publicMD;
 		}
 
-		free() {
-			this.then(_ => {
-				if ((this.handleLocal != undefined) && (this.handleLocal != null)) {
-					safeApp.free(this.handleLocal);
-					console.log('App handle freed: ', this.handleLocal);
-				}
-				this.handleLocal = undefined;
-				this.handlePromise = null;
-				console.log('App.free() completed');
-			});
-		}
 	};
 
 	var MutableData = class extends HandleSynchronisedSafeChild {
 
+		quickSetup(data, name, desc) {
+			return this.then(_ => safeMutableData.quickSetup(this.handle, data, name, desc));
+		}
+
 		getPermissions() {
-			var permissions = new Permissions();
-			permissions.handlePromise = this.handlePromise.then((mdHandle) => safeMutableData.getPermissions(mdHandle));
-			return permissions;
+			return new Permissions(_ => safeMutableData.getPermissions(this.handle), this.root);
```

Comment:
I have a feeling that if we run such a method just after creating a `HandleSynchronisedSafeChild`, we may have `this.handle` being _null_ before `handlePromise` is fulfilled.

Reply:
It won't. There is no forking. However, it isn't obvious why. Let me step through it:

What we've written here, long hand, is:

```
function no_name () {
  return safeMutableData.getPermissions(this.handle);
}
```

That anonymous function gets wrapped into a Promise by `HandleSynchronisedSafeChild` with `root.then(handlePromise)` so that the `no_name()` function above will not call `safeMutableData.getPermissions(this.handle)` _until after_ `root.then()` has caught up with all the previous promises in the chain. Those previous promises should have created and satisfied `this.handle` before `no_name()` is called.